rls/src/test/java/io/grpc/rls/internal/DoNotUseDirectScheduledExecutorService.java

```diff
@@ -0,0 +1,243 @@
+    checkArgument(delay > 0, "delay must be positive");
+    ScheduledRunnable scheduledRunnable =
+        new ScheduledRunnable(currTimeNanos + TimeUnit.NANOSECONDS.convert(delay, unit), command);
+    scheduledCommands.add(scheduledRunnable);
+    return scheduledRunnable.scheduledFuture;
+  }
+
+  FakeTimeProvider getFakeTimeProvider() {
+    maybeInit();
+    return new FakeTimeProvider();
+  }
+
+  void forwardTime(long delta, TimeUnit unit) {
+    maybeInit();
+    checkNotNull(unit, "unit");
+    checkArgument(delta > 0, "delta must be positive");
+    long finalTime = currTimeNanos + unit.toNanos(delta);
+
+    if (repeatedCommand.get() != null) {
+      while (finalTime >= nextRun) {
+        scheduledCommands.add(new ScheduledRunnable(nextRun, repeatedCommand.get()));
+        nextRun += period;
+      }
+    }
+
+    while (!scheduledCommands.isEmpty()
+        && scheduledCommands.peek().scheduledTimeNanos <= finalTime) {
+      ScheduledRunnable scheduledCommand = scheduledCommands.poll();
+      try {
+        // pretend to run at the scheduled time
+        currTimeNanos = scheduledCommand.scheduledTimeNanos;
+        scheduledCommand.run();
+      } catch (Throwable t) {
+        throw new RuntimeException("failed to run scheduled command: " + scheduledCommand, t);
+      }
+    }
+
+    this.currTimeNanos = finalTime;
+  }
+
+  private final class ScheduledRunnable implements Runnable {
+    private final long scheduledTimeNanos;
+    private final Runnable command;
+    private final ScheduledFuture<?> scheduledFuture = new ScheduledRunnable.FakeScheduledFuture();
+    private final Object lock = new Object();
+    private boolean running = false;
+    private boolean done = false;
+
+    public ScheduledRunnable(long scheduledTimeNanos, Runnable command) {
+      this.scheduledTimeNanos = scheduledTimeNanos;
+      this.command = checkNotNull(command, "command");
+    }
+
+    @Override
+    public void run() {
+      synchronized (lock) {
+        if (!scheduledFuture.isCancelled()) {
+          running = true;
+          command.run();
+          done = true;
+        }
+      }
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("scheduledTimeNanos", scheduledTimeNanos)
+          .add("command", command)
+          .add("scheduledFuture", scheduledFuture)
+          .add("running", running)
+          .add("done", done)
+          .toString();
+    }
+
+    private class FakeScheduledFuture implements ScheduledFuture<Object> {
+      boolean cancelled = false;
+
+      @Override
+      public long getDelay(TimeUnit unit) {
+        return unit.convert(scheduledTimeNanos - currTimeNanos, TimeUnit.NANOSECONDS);
+      }
+
+      @Override
+      public int compareTo(Delayed unused) {
+        throw new UnsupportedOperationException();
+      }
+
+      @Override
+      public synchronized boolean cancel(boolean mayInterruptIfRunning) {
```

Comment:
If you really want ThreadSafe, isCancelled() and isDone() also need `synchronized`.

Reply:
yeah removing it makes sense. making it thread safe requires little more work, but it is really meant to be "direct".